griptape/drivers/image_query/openai_vision_image_query_driver.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+from typing import Literal, Optional
+
+from attr import define, field, Factory
+from openai.types.chat import (
+    ChatCompletionUserMessageParam,
+    ChatCompletionContentPartParam,
+    ChatCompletionContentPartTextParam,
+    ChatCompletionContentPartImageParam,
+)
+import openai
+
+from griptape.artifacts import ImageArtifact, TextArtifact
+from griptape.drivers.image_query.base_image_query_driver import BaseImageQueryDriver
+
+
+@define
+class OpenAiVisionImageQueryDriver(BaseImageQueryDriver):
+    model: str = field(default="gpt-4-vision-preview", kw_only=True, metadata={"serializable": True})
+    api_type: str = field(default=openai.api_type, kw_only=True, metadata={"serializable": True})
+    api_version: Optional[str] = field(default=openai.api_version, kw_only=True, metadata={"serializable": True})
+    base_url: str = field(default=None, kw_only=True, metadata={"serializable": True})
+    api_key: Optional[str] = field(default=None, kw_only=True)
+    organization: Optional[str] = field(default=openai.organization, kw_only=True, metadata={"serializable": True})
+    image_quality: Literal["auto", "low", "high"] = field(default="auto", kw_only=True, metadata={"serializable": True})
+    max_tokens: Optional[int] = field(default=None, kw_only=True, metadata={"serializable": True})
+    client: openai.OpenAI = field(
+        default=Factory(
+            lambda self: openai.OpenAI(api_key=self.api_key, base_url=self.base_url, organization=self.organization),
+            takes_self=True,
+        )
+    )
+
+    def try_query(self, query: str, images: list[ImageArtifact]) -> TextArtifact:
+        message_parts: list[ChatCompletionContentPartParam] = [
+            ChatCompletionContentPartTextParam(type="text", text=query)
+        ]
+
+        for image in images:
+            message_parts.append(
+                ChatCompletionContentPartImageParam(
+                    type="image_url",
+                    image_url={"url": f"data:{image.mime_type};base64,{image.base64}", "detail": self.image_quality},
+                )
+            )
+
+        messages = ChatCompletionUserMessageParam(content=message_parts, role="user")
+        params = {"model": self.model, "messages": [messages]}
+
+        if self.max_tokens is not None:
+            params["max_tokens"] = self.max_tokens
```

Comment:
if the user doesn't provide max_tokens, what does OpenAI do? does it have a default?

Reply:
Unfortunate backstory: OpenAI's APIs have an outstanding issue where your account's rate limit is penalized for the max tokens requested, not just the tokens returned. This is why the `OpenAiChatPromptDriver` only provides `max_tokens` only when explicitly provided. Vision, specifically, also has another complicating issue where it will default to a very small number of output tokens (like 16) unless otherwise requested. 

The interaction of both of these issues means there really isn't a great way to define a default here. Requesting a high `max_tokens` value will very quickly consume a typical user's rate limit. The right number will depend on the context.